Implement getArrayForUpdate iteratively

danielschemmel opened the pull request:
<!--
Thank you for contributing to KLEE. We are looking forward to reviewing your PR. However, given the small number of active reviewers and our limited time, it might take a while to do so. We aim to get back to each PR within one month, and often do so within one week.

To review your PR, please add a summary of the proposed changes and ensure all items are fulfilled in the checklist above, by placing an "x" inside each applicable pair of brackets. More details about each item can be found in the [Developer's Guide](https://klee.github.io/docs/developers-guide/#pull-requests).
-->

## Summary: 
This deals with a long-standing `FIXME` in the KLEE codebase where `getArrayForUpdate` is implemented recursively, which could lead to KLEE crashing. This PR transforms the recursive code into an iterative version.

This of course introduces a potential performance penalty by storing the required information in an explicit stack on the heap instead of the implicit function call stack. While I personally don't think the performance penalty is a problem, queries with a small (but non-zero) amount of uncached update nodes will incur additional `malloc`s now.

## Checklist:
- [x] The PR addresses a single issue.  If it can be divided into multiple independent PRs, please do so.
- [x] The PR is divided into a logical sequence of commits OR a single commit is sufficient.
- [x] There are no unnecessary commits (e.g. commits fixing issues in a previous commit in the same PR).
- [x] Each commit has a meaningful message documenting what it does.
- [x] All messages added to the codebase, all comments, as well as commit messages are spellchecked.
- [x] The code is commented OR not applicable/necessary.
- [x] The patch is formatted via clang-format OR not applicable (if explicitly overridden leave unchecked and explain).
- [x] There are test cases for the code you added or modified OR no such test cases are required.

MartinNowack commented on lib/Solver/STPBuilder.cpp:
```diff
@@ -465,26 +465,29 @@ ExprHandle STPBuilder::getInitialRead(const Array *root, unsigned index) {
   if (!un) {
-      return getInitialArray(root);
+    un_expr = getInitialArray(root);
   }
```
```suggestion
  } else {
    un_expr = _arr_hash.lookupUpdateNodeExpr(un, un_expr);
  }
```

We end up with `un_expr` either pointing to the array expr or last relevant cached entry.

danielschemmel replied:
This has already been done as part of the loop condition, where it is necessary to shortcut the full iteration if it is cached. Am I missing something here?

MartinNowack replied:
True, I had a different API in mind.

MartinNowack commented on lib/Solver/STPBuilder.cpp:
```diff
@@ -465,26 +465,29 @@ ExprHandle STPBuilder::getInitialRead(const Array *root, unsigned index) {
+    --i;
+    un = update_nodes[i];
+
+    un_expr = vc_writeExpr(vc, getArrayForUpdate(root, un->next.get()),
```
This should be now just referencing the last expr.
```suggestion
    un_expr = vc_writeExpr(vc, un_expr,
```
This way, we get rid of the recursive `getArrayForUpdate` call.

The reference counting can sometimes be a bit tricky but hopefully ASAN will tell ;)

danielschemmel replied:
Yep, that was kind of the point of this PR... Good catch!

MartinNowack commented on lib/Solver/STPBuilder.cpp:
```diff
@@ -465,26 +465,29 @@ ExprHandle STPBuilder::getInitialRead(const Array *root, unsigned index) {
-      }
-      
-      return un_expr;
+  for (std::size_t i = update_nodes.size(); i > 0;) {
```
Maybe use an iterator approach here with `update_nodes.rbegin()`. I would love to have a reversed range-based for-loop in such cases.

danielschemmel replied:
That is available [in C++20](https://en.cppreference.com/w/cpp/ranges/reverse_view) or slightly less pretty [using LLVM](https://llvm.org/doxygen/classllvm_1_1iterator__range.html). Since we are still two C++ versions behind, I used the LLVM variant here.

MartinNowack reviewed (changes requested):
Thanks @danielschemmel for taking care of this. It already looks quite good!

I added some comments only for the STP solver - the other solvers should be handled similarly.

In addition, please add some comments to the code, so we get it more accessible for others as well.

ccadar commented:
@danielschemmel I notice there are many uncovered lines in `STPBuilder.cpp`.  Could you try to change some tests to also use Z3 as a constraint solver, in order to exercise that code?

danielschemmel commented:
> @danielschemmel I notice there are many uncovered lines in `STPBuilder.cpp`. Could you try to change some tests to also use Z3 as a constraint solver, in order to exercise that code?

I had a quick look at how the tests choose the constraint solver, and it turns out most of them simply don't specify the solver backend at all, meaning they will end up picking the default backend. To be precise, there is exactly 1 test that specifically uses STP (test/Solver/STPDumpDebugQueries.c) and two tests thet specifically use Z3 (test/Solver/Z3ConstantArray.c and test/Solver/Z3LargeConstantArray.kquery). No tests explicitly use MetaSMT.

If I read .github/workflows/build.yaml right, then the coverage build is done with the default solver setup of `SOLVERS: STP:Z3`, which leads to STP being the default solver, and therefore the solver used for almost all tests. This config also explains how metasmt seems so well tested: since the coverage check is done with `SOLVERS: STP:Z3`, metasmt is disabled, and therefore has no lines to be covered, while Z3 is enabled but not covered.

This is, however, not quite the whole story, as the test matrix includes a variety of other solver configurations, for example, the "Z3 only" config does just what its name suggests. Therefore, the Z3 code is tested roughly as well as the STP code, but not while reporting coverage. The metaSMT/STP and metaSMT/boolector solvers are similarly well tested, also not while reporting coverage, but in a manner that does not show them as uncovered lines.

A simple and reasonable suggestion is to just bluntly turn down the solver selection in the coverage run to only include one single solver which would then be the default solver and used in the tests.

ccadar commented:
@danielschemmel Good analysis, thanks! 
Indeed, the reported coverage is misleading, as you say.  So the easy solution would be to enable only STP in the coverage configuration.  The better one would be to run the tests once with STP, once with Z3, and once with metaSMT.  However, the coverage in the Z3 and metaSMT builders are likely to mirror the one in STP, so perhaps the first solution is good enough.  
Do you want to add an extra commit here that only runs the Coverage target with STP and make sure it works as expected?

danielschemmel commented:
@ccadar We *are* running the tests with z3 and metasmt (in two different variants even), we just don't collect coverage while doing so.

I will add the easy fix now, I can figure out something that captures the actual coverage better later on, if you would like me to. This problem exists similarly for different LLVM versions.

danielschemmel commented:
It seems the change had the expected effect.

ccadar commented:
@danielschemmel Of course, we run many additional configurations across runs.  And the numbers are strictly for one configuration.  As I argued above, I think using only STP there is fine.  I'll wait for others to comment, after which we can merge this.  CC: @MartinNowack, @251

ccadar commented:
@danielschemmel If we go for #1531, then we should drop the second commit here.

MartinNowack commented:
Nice changes! I agree with the comment from @ccadar , we should fix the coverage issue in the other PR.

The only downside we have now is that we iterate over all update nodes, make a copy of the reference to iterate over all the notes again. That might be a bit costly. On the other hand, that might be less of an overall problem. 

Can you add some comments to the code, so we get it more accessible for others as well, i.e. adding high-level ideas of what the code does.

251 commented on lib/Solver/Z3Builder.cpp:
```diff
@@ -427,22 +428,25 @@ Z3ASTHandle Z3Builder::getInitialRead(const Array *root, unsigned index) {
+  std::vector<const UpdateNode *> update_nodes;
+  for (; un && !_arr_hash.lookupUpdateNodeExpr(un, un_expr);
+       un = un->next.get()) {
+    update_nodes.emplace_back(un);
```
What's the point of `emplace_back` instead of `push_back` here?

MartinNowack replied:
A good habit ;)

danielschemmel replied:
If you prefer `push_back`, that is an easy enough change...

danielschemmel commented:
I have replaced `emplace_back` with `push_back`, as Frank seems to prefer it here quite a bit, dropped the commit restricting ourselves to stp in the coverage check, and added some comments.

Personally, I don't really like the comments that came out of it, as this function is really just a caching `map`.

MartinNowack commented on lib/Solver/MetaSMTBuilder.h:
```diff
@@ -172,22 +173,34 @@ template <typename SolverContext>
 MetaSMTBuilder<SolverContext>::getArrayForUpdate(const Array *root,
                                                  const UpdateNode *un) {
-
+  // Iterate over the update nodes, until we find a cached version of the array,
```
```suggestion
  // Iterate over the update nodes, until we find a cached version of the node,
```

danielschemmel replied:
Done

MartinNowack commented on lib/Solver/MetaSMTBuilder.h:
```diff
@@ -172,22 +173,34 @@ template <typename SolverContext>
+  // `un_expr` now holds an expression for the array - either from cache or by
+  // virtue of being the initial array expression
+
+  // now we can apply create the expression for the current version of the array
```
```suggestion
// Create and cache solver expressions based on the update nodes starting from the oldest.
```

MartinNowack replied:
And remove the other two lines

danielschemmel replied:
Done

MartinNowack reviewed (commented):
@danielschemmel Almost there. I added some suggestions.
